src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt

```diff
@@ -0,0 +1,308 @@
+import org.elasticsearch.action.support.ActionFilter
+import org.elasticsearch.action.support.ActionFilterChain
+import org.elasticsearch.action.support.IndicesOptions
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.common.Strings
+import org.elasticsearch.common.xcontent.DeprecationHandler
+import org.elasticsearch.common.xcontent.NamedXContentRegistry
+import org.elasticsearch.common.xcontent.XContentFactory
+import org.elasticsearch.common.xcontent.json.JsonXContent
+import org.elasticsearch.tasks.Task
+
+private val logger = LogManager.getLogger(FieldCapsFilter::class.java)
+
+@Suppress("UNCHECKED_CAST", "SpreadOperator", "TooManyFunctions")
+class FieldCapsFilter(
+    val clusterService: ClusterService,
+    val indexNameExpressionResolver: IndexNameExpressionResolver
+) : ActionFilter {
+
+    override fun <Request : ActionRequest?, Response : ActionResponse?> apply(
+        task: Task,
+        action: String,
+        request: Request,
+        listener: ActionListener<Response>,
+        chain: ActionFilterChain<Request, Response>
+    ) {
+        if (request is FieldCapabilitiesRequest) {
+            val indices = request.indices().map { it.toString() }.toTypedArray()
+            val concreteIndices = indexNameExpressionResolver.concreteIndexNames(clusterService.state(), request.indicesOptions(), *indices)
+            val rollupIndices = mutableSetOf<String>()
+            val filteredIndices = mutableSetOf<String>()
+            for (index in concreteIndices) {
+                val isRollupIndex = RollupSettings.ROLLUP_INDEX.get(clusterService.state().metadata.index(index).settings)
+                if (isRollupIndex) {
+                    rollupIndices.add(index)
+                } else {
+                    filteredIndices.add(index)
+                }
+            }
+
+            val hasRollupIndices = rollupIndices.size > 0
+            if (hasRollupIndices) {
+                request.indices(*filteredIndices.toTypedArray())
+            }
+
+            val shouldShortCircuit = rollupIndices.size > 0 && filteredIndices.size == 0
+            if (shouldShortCircuit) {
+                val rewrittenResponse = rewriteResponse(mapOf(), arrayOf(), rollupIndices)
+                listener.onResponse(rewrittenResponse as Response)
+            } else {
+                chain.proceed(task, action, request, object : ActionListener<Response> {
+                    override fun onResponse(response: Response) {
+                        if (hasRollupIndices) {
+                            logger.info("Has rollup indices will rewrite field caps response")
+                            response as FieldCapabilitiesResponse
+                            val reWrittenResponse = rewriteResponse(response.get(), response.indices, rollupIndices)
+                            listener.onResponse(reWrittenResponse as Response)
```

Comment:
nit: Here it's reWritten and above it's rewritten

Reply:
makes sense, will update